packages/nitro-protocol/contracts/MultiAssetHolder.sol

```diff
@@ -2,104 +2,220 @@
+        uint256 amount
+    ) external override payable {
+        require(!_isExternalDestination(destination), 'Deposit to external destination');
+        uint256 amountDeposited;
+        // this allows participants to reduce the wait between deposits, while protecting them from losing funds by depositing too early. Specifically it protects against the scenario:
+        // 1. Participant A deposits
+        // 2. Participant B sees A's deposit, which means it is now safe for them to deposit
+        // 3. Participant B submits their deposit
+        // 4. The chain re-orgs, leaving B's deposit in the chain but not A's
+        require(holdings[asset][destination] >= expectedHeld, 'holdings < expectedHeld');
+        require(
+            holdings[asset][destination] < expectedHeld.add(amount),
+            'holdings already sufficient'
+        );
+
+        // The depositor wishes to increase the holdings against channelId to amount + expectedHeld
+        // The depositor need only deposit (at most) amount + (expectedHeld - holdings) (the term in parentheses is non-positive)
+
+        amountDeposited = expectedHeld.add(amount).sub(holdings[asset][destination]); // strictly positive
+        // require successful deposit before updating holdings (protect against reentrancy)
+        if (asset == address(0)) {
+            require(msg.value == amount, 'Incorrect msg.value for deposit');
+        } else {
+            // require successful deposit before updating holdings (protect against reentrancy)
+            require(
+                IERC20(asset).transferFrom(msg.sender, address(this), amountDeposited),
+                'Could not deposit ERC20s'
+            );
+        }
+
+        holdings[asset][destination] = holdings[asset][destination].add(amountDeposited);
+        emit Deposited(destination, asset, amountDeposited, holdings[asset][destination]);
+
+        if (asset == address(0)) {
+            // refund whatever wasn't deposited.
+            uint256 refund = amount.sub(amountDeposited);
+            (bool success, ) = msg.sender.call{value: refund}(''); //solhint-disable-line avoid-low-level-calls
+            require(success, 'Could not refund excess funds');
+        }
+    }
+
     /**
      * @notice Transfers as many funds escrowed against `channelId` as can be afforded for a specific destination. Assumes no repeated entries.
      * @dev Transfers as many funds escrowed against `channelId` as can be afforded for a specific destination. Assumes no repeated entries.
      * @param fromChannelId Unique identifier for state channel to transfer funds *from*.
-     * @param allocationBytes The abi.encode of AssetOutcome.Allocation
+     * @param outcomeBytes The encoded Outcome of this state channel
      * @param indices Array with each entry denoting the index of a destination to transfer funds to. An empty array indicates "all".
      */
     function transfer(
+        uint256 assetIndex, // TODO consider a uint48?
```

Comment:
That's an interesting suggestion, and is relevant over in the exit-format repo. 

I think a dictionary might be easier to work with, but I don't think it is possible to implement, sadly. 

In[ typescript](https://www.typescriptlang.org/docs/handbook/utility-types.html#recordkeystype), I think your idea would look something like this:

```ts
type Outcome = Record<Address, AssetOutcome>
````
In solidity we can use a [mapping](https://docs.soliditylang.org/en/v0.8.6/types.html#mapping-types) to get something similar, but that only works for `storage` variable. We would then be using a separate storage slot for each `AssetOutcome`, which is actually the very thing we are trying to avoid doing here. It's also not clear how we can encode a mapping, which is something we need (in order to store the hash, etc). 

The only other thing we could try to do is have:

```solidity
struct Outcome {
    AssetOutcome 0x000;
    AssetOutcome 0xabc;
}
```
but this doesn't work either, for a couple of reasons. The keys are invalid (addresses are not valid identifiers), and they have to be specified at compile-time (we need them to be specified at runtime).

Reply:
That makes sense. Thanks for the enlightening explanation!